File: windows/SharedPrinterModule/SharedPrinterModule.cs
using System;
using System.Collections.Generic;
using System.Drawing.Printing;
using System.IO.Ports;
using System.Linq;

namespace SharedPrinterModule
{
    public class SharedPrinterModule
    {
        private readonly RawPrinterClient rawPrinterClient = new RawPrinterClient();
        private readonly SerialPrinterClient serialPrinterClient = new SerialPrinterClient();
        private Dictionary<string, object> connectedPrinter;

        public void Initialize()
        {
            // Kept for parity with the SDK bridge.
        }

        public IList<Dictionary<string, object>> DiscoverPrinters()
        {
            var printers = new List<Dictionary<string, object>>();

            printers.AddRange(
                PrinterSettings.InstalledPrinters
                    .Cast<string>()
                    .Select(MapQueuePrinter));

            printers.AddRange(
                SerialPort.GetPortNames()
                    .OrderBy(portName => portName)
                    .Select(MapSerialPrinter));

            return printers;
        }

        public Dictionary<string, object> Connect(string printerId)
        {
            var printer = DiscoverPrinters().FirstOrDefault(candidate =>
                string.Equals(GetValue(candidate, "id"), printerId, StringComparison.Ordinal));

            if (printer == null)
            {
                throw new InvalidOperationException(string.Format("Printer not found: {0}", printerId));
            }

            connectedPrinter = printer;
            return printer;
        }

        public void Disconnect(string printerId)
        {
            if (connectedPrinter == null)
            {
                return;
            }

            if (printerId == null ||
                string.Equals(GetValue(connectedPrinter, "id"), printerId, StringComparison.Ordinal))
            {
                connectedPrinter = null;
            }
        }

        public Dictionary<string, object> GetConnectedPrinter()
        {
            return connectedPrinter;
        }

        public Dictionary<string, object> GetStatus(string printerId)
        {
            var isConnected = connectedPrinter != null &&
                (printerId == null ||
                 string.Equals(GetValue(connectedPrinter, "id"), printerId, StringComparison.Ordinal));

            return new Dictionary<string, object>
            {
                { "state", isConnected ? "connected" : "disconnected" },
                { "updatedAt", DateTimeOffset.UtcNow.ToString("O") },
                { "isOffline", !isConnected }
            };
        }

        public Dictionary<string, object> PrintReceipt(string jobId, string printerId, IList<IList<int>> chunks)
        {
            var printer = RequireConnectedPrinter(printerId);
            WriteToConnectedPrinter(printer, FlattenChunks(chunks), "Receipt-" + jobId);

            return new Dictionary<string, object>
            {
                { "status", "completed" },
                { "message", string.Format("Receipt job {0} sent to {1}.", jobId, GetValue(printer, "name")) }
            };
        }

        public Dictionary<string, object> PrintLabel(string jobId, string printerId, IList<IList<int>> chunks)
        {
            var printer = RequireConnectedPrinter(printerId);
            WriteToConnectedPrinter(printer, FlattenChunks(chunks), "Label-" + jobId);

            return new Dictionary<string, object>
            {
                { "status", "completed" },
                { "message", string.Format("Label job {0} sent to {1}.", jobId, GetValue(printer, "name")) }
            };
        }

        private Dictionary<string, object> RequireConnectedPrinter(string printerId)
        {
            if (connectedPrinter == null ||
                !string.Equals(GetValue(connectedPrinter, "id"), printerId, StringComparison.Ordinal))
            {
                throw new InvalidOperationException(string.Format("Printer is not connected: {0}", printerId));
            }

            return connectedPrinter;
        }

        private void WriteToConnectedPrinter(Dictionary<string, object> printer, byte[] bytes, string documentName)
        {
            var transportType = GetValue(printer, "transportType");
            var address = GetValue(printer, "address");
            var name = GetValue(printer, "name");

            if (string.Equals(transportType, "serial", StringComparison.OrdinalIgnoreCase))
            {
                serialPrinterClient.Write(address, bytes);
                return;
            }

            rawPrinterClient.Write(name, bytes, documentName);
        }

        private Dictionary<string, object> MapQueuePrinter(string printerName)
        {
            var settings = new PrinterSettings { PrinterName = printerName };
            var transportType = InferQueueTransportType(printerName);

            return new Dictionary<string, object>
            {
                { "id", printerName },
                { "name", printerName },
                { "transportType", transportType },
                { "address", printerName },
                { "isDefault", settings.IsDefaultPrinter },
                { "isValid", settings.IsValid }
            };
        }

        private Dictionary<string, object> MapSerialPrinter(string portName)
        {
            return new Dictionary<string, object>
            {
                { "id", portName },
                { "name", string.Format("Serial Printer ({0})", portName) },
                { "transportType", "serial" },
                { "address", portName },
                { "isDefault", false },
                { "isValid", true }
            };
        }

        private string InferQueueTransportType(string portName)
        {
            if (portName.StartsWith("USB", StringComparison.OrdinalIgnoreCase))
            {
                return "usb";
            }

            if (portName.StartsWith("COM", StringComparison.OrdinalIgnoreCase))
            {
                return "serial";
            }

            if (portName.StartsWith("IP_", StringComparison.OrdinalIgnoreCase) ||
                portName.Contains("."))
            {
                return "wifi";
            }

            return "socket";
        }

        private byte[] FlattenChunks(IList<IList<int>> chunks)
        {
            var bytes = new List<byte>();
            foreach (var chunk in chunks)
            {
                bytes.AddRange(chunk.Select(value => (byte)(value & 0xFF)));
            }

            return bytes.ToArray();
        }

        private string GetValue(Dictionary<string, object> printer, string key)
        {
            object value;
            if (!printer.TryGetValue(key, out value) || value == null)
            {
                return string.Empty;
            }

            return value.ToString();
        }
    }
}

File: android/src/main/java/com/nukkad/sharedprinter/SharedPrinterModule.java
package com.nukkad.sharedprinter;

import android.hardware.usb.UsbConstants;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbManager;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SharedPrinterModule {
    private static final int USB_WRITE_TIMEOUT_MS = 4_000;

    private final AndroidUsbPrinterConnection usbPrinterConnection;

    public SharedPrinterModule(UsbManager usbManager) {
        this.usbPrinterConnection = new AndroidUsbPrinterConnection(usbManager);
    }

    public void initialize() {
        // Kept for API symmetry with the TypeScript bridge.
    }

    public List<Map<String, Object>> discoverPrinters() {
        List<Map<String, Object>> printers = new ArrayList<>();

        for (UsbDevice device : usbPrinterConnection.discoverDevices().values()) {
            if (!isPotentialPrinter(device)) {
                continue;
            }

            printers.add(mapPrinter(device));
        }

        return printers;
    }

    public Map<String, Object> connect(String printerId) {
        UsbDevice device = findDevice(printerId);
        if (device == null) {
            throw new IllegalArgumentException("Printer not found: " + printerId);
        }

        if (!usbPrinterConnection.hasPermission(device)) {
            throw new IllegalStateException("USB permission missing for printer: " + printerId);
        }

        usbPrinterConnection.connect(device);
        return mapPrinter(device);
    }

    public void disconnect(String printerId) {
        if (printerId == null || usbPrinterConnection.isConnected(printerId)) {
            usbPrinterConnection.disconnect();
        }
    }

    public Map<String, Object> getConnectedPrinter() {
        UsbDevice connectedDevice = usbPrinterConnection.getConnectedDevice();
        return connectedDevice == null ? null : mapPrinter(connectedDevice);
    }

    public Map<String, Object> getStatus(String printerId) {
        boolean isConnected = printerId == null
                ? usbPrinterConnection.getConnectedDevice() != null
                : usbPrinterConnection.isConnected(printerId);

        HashMap<String, Object> status = new HashMap<>();
        status.put("state", isConnected ? "connected" : "disconnected");
        status.put("updatedAt", Instant.now().toString());
        status.put("isOffline", !isConnected);
        return status;
    }

    public Map<String, Object> printReceipt(String jobId, String printerId, List<List<Integer>> chunks) {
        writeChunks(printerId, chunks);
        return buildPrintResult(jobId, "Receipt printed successfully.");
    }

    public Map<String, Object> printLabel(String jobId, String printerId, List<List<Integer>> chunks) {
        writeChunks(printerId, chunks);
        return buildPrintResult(jobId, "Label printed successfully.");
    }

    private void writeChunks(String printerId, List<List<Integer>> chunks) {
        if (!usbPrinterConnection.isConnected(printerId)) {
            throw new IllegalStateException("Printer is not connected: " + printerId);
        }

        for (List<Integer> chunk : chunks) {
            usbPrinterConnection.write(toByteArray(chunk), USB_WRITE_TIMEOUT_MS);
        }
    }

    private UsbDevice findDevice(String printerId) {
        for (UsbDevice device : usbPrinterConnection.discoverDevices().values()) {
            if (device.getDeviceName().equals(printerId)) {
                return device;
            }
        }

        return null;
    }

    private boolean isPotentialPrinter(UsbDevice device) {
        return device.getDeviceClass() == UsbConstants.USB_CLASS_PRINTER
                || device.getInterfaceCount() > 0;
    }

    private Map<String, Object> mapPrinter(UsbDevice device) {
        HashMap<String, Object> printer = new HashMap<>();
        printer.put("id", device.getDeviceName());
        printer.put("name", device.getProductName() == null ? "USB Printer" : device.getProductName());
        printer.put("transportType", "usb");
        printer.put("vendorId", String.format("0x%04X", device.getVendorId()));
        printer.put("productId", String.format("0x%04X", device.getProductId()));
        printer.put("address", device.getDeviceName());
        printer.put("hasPermission", usbPrinterConnection.hasPermission(device));
        return printer;
    }

    private Map<String, Object> buildPrintResult(String jobId, String message) {
        HashMap<String, Object> result = new HashMap<>();
        result.put("jobId", jobId);
        result.put("status", "completed");
        result.put("message", message);
        return result;
    }

    private byte[] toByteArray(List<Integer> values) {
        byte[] bytes = new byte[values.size()];
        for (int index = 0; index < values.size(); index++) {
            bytes[index] = (byte) (values.get(index) & 0xFF);
        }
        return bytes;
    }
}
